cmd/heal.go

```diff
@@ -0,0 +1,359 @@
+}
+
+// newHealSys initializes healing
+func newHealSys(dir string) *HealSys {
+	if dir == "" {
+		return nil
+	}
+	sys := &HealSys{Dir: dir, errorCh: make(chan ErrorLog, 10000), healCh: make(chan ErrorLog, 10000)}
+	go sys.heal()
+	return sys
+}
+
+func (sys *HealSys) init() {
+	if sys == nil {
+		return
+	}
+	if err := os.MkdirAll(sys.Dir, 0777); err != nil {
+		logger.FatalIf(err, fmt.Sprintf("Unable to initialize '%s' dir", sys.Dir))
+	}
+	files, err := ioutil.ReadDir(sys.Dir)
+	if err != nil {
+		logger.FatalIf(err, fmt.Sprintf("Invalid error log dir %s", sys.Dir))
+	}
+	for _, fi := range files {
+		sys.process(fi.Name())
+	}
+	go sys.startErrLogger()
+}
+
+// process sends error logs to the heal channel for an attempt to heal.
+func (sys *HealSys) process(fname string) {
+	fpath := pathJoin(sys.Dir, fname)
+	file, err := os.Open(fpath)
+	if err != nil {
+		logger.LogIf(GlobalContext, err)
+		return
+	}
+	defer file.Close()
+	bio := bufio.NewScanner(file)
+	for bio.Scan() {
+		var errLog ErrorLog
+		if err = json.Unmarshal(bio.Bytes(), &errLog); err != nil {
+			// TODO: decide on retry
+			continue
+		}
+		sys.healCh <- errLog
+	}
+	if err = bio.Err(); err != nil {
+		logger.LogIf(GlobalContext, err)
+	}
+	os.Remove(fpath)
+}
+
+// send error log to error channel
+func (sys *HealSys) send(log ErrorLog) {
+	if sys == nil {
+		return
+	}
+	select {
+	case sys.errorCh <- log:
+		return
+	case <-GlobalContext.Done():
+		return
+	}
+}
+
+// startErrLogger logs radio API call errors to a temporary file for further processing
+// by healing routine.
+func (sys *HealSys) startErrLogger() {
+	prevLogName := ""
+	currLogName := errLogName()
+	fptr, err := newErrorFile(pathJoin(sys.Dir, currLogName))
+	if err != nil {
+		logger.LogIf(GlobalContext, err)
+	}
+	ticker := time.NewTicker(30 * time.Second)
+	defer ticker.Stop()
+	for {
+		select {
+		case log, ok := <-sys.errorCh:
+			if !ok {
+				continue
+			}
+			for {
+				jsonBytes, err := json.Marshal(log)
+				if err != nil {
+					logger.LogIf(GlobalContext, err)
+				}
+				if _, err = fptr.Write(jsonBytes); err != nil {
+					logger.LogIf(GlobalContext, err)
+					continue
+				}
+				fptr.WriteString("\n")
+				break
+			}
+		case <-ticker.C:
+			fptr.Close()
+			prevLogName = currLogName
+			currLogName = errLogName()
+			fptr, err = newErrorFile(pathJoin(sys.Dir, currLogName))
+			if err != nil {
+				logger.LogIf(GlobalContext, err)
+			}
+			// go heal
+			sys.process(prevLogName)
+		case <-GlobalContext.Done():
+			fptr.Close()
+			return
+		}
+	}
+}
+
+// heal() retries failed mutable API operations and attempts to
+// get the object in sync between mirrored clients.
+func (sys *HealSys) heal() {
+	l, ok := globalObjectAPI.(*radioObjects)
+	if !ok {
+		logger.LogOnceIf(GlobalContext, fmt.Errorf(""), "")
+		return
+	}
+	for {
+		select {
+		case log, ok := <-sys.healCh:
+			if !ok {
+				continue
+			}
+			switch log.Op {
+			case opPutObject:
+				l.healPutObject(log)
```

Comment:
implement public APIs such that it can be used by erasure coding as well when implemented @poornas avoids unnecessary reflection code.